src/library/scala/concurrent/impl/ExecutionContextImpl.scala

```diff
@@ -53,39 +52,55 @@ private[concurrent] object ExecutionContextImpl {
       thread
     }
 
-    // As per ThreadFactory contract newThread should return `null` if cannot create new thread.
-    def newThread(runnable: Runnable): Thread =
-      if (reserveThread())
-        wire(new Thread(new Runnable {
-          // We have to decrement the current thread count when the thread exits
-          override def run() = try runnable.run() finally deregisterThread()
-        })) else null
-
-    def newThread(fjp: ForkJoinPool): ForkJoinWorkerThread =
-      if (reserveThread()) {
-        wire(new ForkJoinWorkerThread(fjp) with BlockContext {
-          // We have to decrement the current thread count when the thread exits
-          final override def onTermination(exception: Throwable): Unit = deregisterThread()
-          final override def blockOn[T](thunk: =>T)(implicit permission: CanAwait): T = {
-            var result: T = null.asInstanceOf[T]
-            ForkJoinPool.managedBlock(new ForkJoinPool.ManagedBlocker {
-              @volatile var isdone = false
-              override def block(): Boolean = {
-                result = try {
-                    // When we block, switch out the BlockContext temporarily so that nested blocking does not created N new Threads
-                    BlockContext.withBlockContext(BlockContext.defaultBlockContext) { thunk }
-                  } finally {
-                    isdone = true
-                  }
-
-                true
-              }
-              override def isReleasable = isdone
-            })
-            result
+    private[this] final def allocateNormalThread(): Boolean =
+      Thread.currentThread match {
+        case w: WorkerThread if w.isBlocking => false
+        case _ => true
+      }
+
+    override def newThread(runnable: Runnable): Thread =
+      if (allocateNormalThread()) wire(new Thread(runnable))
+      else if (reserveExtraThread()) wire(new Thread(new Runnable {
+          def run(): Unit = try runnable.run() finally deregisterExtraThread()
+        }))
+      else null
+
+    override def newThread(fjp: ForkJoinPool): ForkJoinWorkerThread =
+      if (allocateNormalThread()) wire(new WorkerThread(fjp))
+      else if (reserveExtraThread()) wire(new ExtraThread(fjp))
+      else null
+
+      final class ExtraThread(fjp: ForkJoinPool) extends ForkJoinWorkerThread(fjp) with BlockContext {
+        final override def onTermination(exception: Throwable): Unit = deregisterExtraThread()
+        final override def blockOn[T](thunk: =>T)(implicit permission: CanAwait): T =
+          BlockContext.defaultBlockContext.blockOn(thunk)
+      }
+
+      final class WorkerThread(fjp: ForkJoinPool) extends ForkJoinWorkerThread(fjp) with BlockContext {
+        @volatile var isBlocking: Boolean = false
+        final override def blockOn[T](thunk: =>T)(implicit permission: CanAwait): T = if (isBlocking) {
```

Comment:
Also, just an observation. I noticed that scala.concurrent.blocking has got only effect when global thread pool is used as an executor context. When a client implements (or sources 3rd party) another executor context, there is no way to instruct executor context about blocking call. As a result, the only correctly working with `blocking` executor context is the global one (or the one created from `null` executor). I see it as incomplete API case or unfair favoring for scala internal implementation of an executor context.

Reply:
@avkonst No, anyone who writes an ExecutionContext, and anyone who wraps one, and anyone who wants to install some other BlockContext within a scope is free to do so. So it is most definitely not favoring any implementation. See: http://viktorklang.com/blog/Futures-in-Scala-protips-4.html